content/realtime/channel-params.textile

```diff
@@ -0,0 +1,91 @@
+
+h2(#rewind). rewind parameter
+
+The @rewind@ channel parameter relates to the initial attachment of a connection to a channel, and expresses the intent to attach to the channel at a position, or a point in time, in the past (that is, effectively "rewinding" the channel for the purposes of the present attachment).
+
+A @rewind@ parameter can express a channel position in terms of a number of messages, or a time interval.
+
+A @rewind@ value that is simply a number @n@ (eg @rewind=1@) is a request to attach to the channel at a position @n@ messages before the present position. If that attachment is successful, and one or more messages exist on the channel prior to the present position, then those messages will be delivered to the subscriber immediately after the attachment has completed, and before any subsequent messages that arise in real time.
+
+If fewer than the requested number of messages exists on the channel (including the case that there are no prior messages), then the available messages are sent; this does not constitute an error.
+
+A @rewind@ value can also be a string that is a time interval specifier. Supported specifier values express an integral number of seconds (eg @15s@), minutes (eg @5m@) or hours (eg @4h@). If that attachment is successful, and one or more messages exist on the channel in the given time interval prior to the present time, then those messages will be delivered to the subscriber immediately after the attachment has completed, and before any subsequent messages that arise in real time.
+
+If fewer than the requested number of messages exists on the channel in that interval (including the case that there are no messages), then the available messages are sent; this does not constitute an error.
+
+If the number of messages within the specified interval is greater than the attachment @limit@ (which defaults to 100), then only the most recent messages up to that limit are sent. The attachment succeeds, but truncation of the message backlog is indicated as a non-fatal error in the attachment response.
+
+It is also possible to specify a different limit via a @limit@ channel parameter. So, for example, to request up to 10 messages, in a window 5m before the present time, specify a channel parameter string of @[?rewind=5m&limit=10]@.
+
+By default, a maximum of two minutes of channel history is available when attaching. This means that a rewind time specifier of greater than two minutes will only be able to rewind by two minutes. If a channel has persistence enabled, then it is possible to rewind back in time by up to the persistence TTL on the channel. 
+
+The channel position expressed by a @rewind@ parameter has an effect only on an initial channel attachment. Any subsequent reattachment of the same channel on the same connection, in order to resume the connection, will attempt to resume with continuity from the point at which the connection dropped.
```

Comment:
What happens if the re-attachment cannot provide continuity i.e. connection has gone from suspended to connected?

Reply:
If continuity is not possible, then the response to the attach request indicates the loss of continuity and the reason as a non-fatal error, in the usual way.

Reply:
> then the response to the attach request indicates the loss of continuity and the reason as a non-fatal error, in the usual way.

Ok, so then for my understanding, are we saying that:

* Customers should listen for channel continuity failures, and use the history API if they want to recover messages at that point, as usual.
* The client SDK knows which channel attributes should be sent or not when calling ATTACH because the realtime system may in fact not know which ATTACH requests are as a result of state lost vs a new attach.  For example, user connects, attaches to `foo`, then becomes disconnected for 10 minutes, resumes (and fails), so automatic re-ATTACH is sent on `foo` without the `rewind` param, whilst the user happens to call attach on `bar` which includes rewind, and the `rewind` param is sent at that time.  If this is the case, this introduces some complexity into the client SDK and means arbitrary key/val pairs are no longer arbitrary.  Perhaps if we have to have options that mean different things on reattach, they could be suffixed with `!` to mean always send for example (not suggesting that is the solutin, just an example).

Reply:
> mean different things on reattach, they could be suffixed with ! to mean always send for example

I'd be against adding that complexity to the client libs (definitely when using channel-paramterers-in-the-name, but also when using client libs that support them natively).

In the case of a resume attempt that fails, realtime knows that that's the case and can itself decide to not apply the rewind option. So the only problem case is a clean attachment as a result of the client lib removing stale connection state after two minutes -- but in that case, is it actually a problem having the rewind param apply again? As long as time-based rewind is only live history, then it's only the last two minutes meaning that it's a subset of the messages the user would have missed. That seems like perfectly reasonable behaviour.

Admittedly if rewind can silently reach into persisted history with time-based options (eg the user can specify rewind=1h) then it becomes more of a problem. But I didn't realised we'd decided to do that -- see other comment.

Reply:
> But in that case, is it actually a problem having the rewind param apply again? 

Well potentially yes because it's a change in behaviour.  I don't like telling customers rewind only applies once, except when the client decides to drop the state and you don't know when that happens, so just don't rely on this behaviour.

> As long as time-based rewind is only live history, then it's only the last two minutes meaning that it's a subset of the messages the user would have missed. That seems like perfectly reasonable behaviour.

I think it's unreasonable to state sometimes one thing happens, and sometimes something else happen.  

I think a channel attach flag should be included now from client libs to indicate that the attach is a reattach vs an explicit attach with params.  This way realtime can always decide on the right behaviour, and could, for example treat different params differently without having to bake in any logic into the SDKs.

Reply:
@mattheworiordan discussed this with @paddybyers just now. Conclusions:
- In most cases when the user specifies, say, rewind=2m and then does a reattach after eg 3 minutes due to internet connection loss, the most useful behaviour is probably to give them the last 2 minutes of messages, rather than nothing
- We should have an api that both works for users now and can nicely transition to a future when we can do arbitrary time resumes (whether with liftbridge or our current persisted history)
- The case that I was worrying about (rewind=72h, reattach after three minutes, get 72h of history) I should not worry about: any future in which we allow more than rewind=2m / rewind=1 (fulfilled from persisted history using timeserials etc) is a future in which the client library would not clear its connection state after 2 minutes, so realtime would know it's a resume and be able to fulfil it from the requested resume point.
- So we think an API of the form 'the rewind param applies whenever the channel could not resume from its previous point' is probably the best option right now: easy to explain, doesn't require any rewind-specific client library changes, easy to extend to a future where we can resume a lot more than we do now, but also works well for what we can do now (ie mostly backfill from live history)

Reply:
> In most cases when the user specifies, say, rewind=2m and then does a reattach after eg 3 minutes due to internet connection loss, the most useful behaviour is probably to give them the last 2 minutes of messages, rather than nothing

I disagree.  If a user loses continuity on a channel, we have a very well documented approach and set of recommendations about letting the developer decide what is most appropriate.  It is not our job to decide that, and the reason we provide that. See https://www.ably.io/documentation/realtime/channels#nonfatal-errors and specifically "If channel continuity is lost in the process, the library will notify you though a resumed flag in the ATTACHED or UPDATE event, so you can decide how to handle that failure.".  This approach contravenes that principle.

> We should have an api that both works for users now and can nicely transition to a future when we can do arbitrary time resumes (whether with liftbridge or our current persisted history)

Agreed

> The case that I was worrying about (rewind=72h, reattach after three minutes, get 72h of history) I should not worry about: any future in which we allow more than rewind=2m / rewind=1 (fulfilled from persisted history using timeserials etc) is a future in which the client library would not clear its connection state after 2 minutes, so realtime would know it's a resume and be able to fulfil it from the requested resume point.

Ok, but there will still be a connection state / channel persistence history limit.  There will always be a limit to how far back one can go.

> So we think an API of the form 'the rewind param applies whenever the channel could not resume from its previous point' is probably the best option right now: 

> easy to explain
No, it's not.

> , doesn't require any rewind-specific client library changes,
We are making changes now.  They don't need to be rewind specific, they are reattach specific so that realtime can decide now and in the future for any other param we may choose to support.  We're making a change now in the spec that could help us in future in treating new undefined params differently based on the explicit vs implicit attach in future.

> easy to extend to a future where we can resume a lot more than we do now
As above, there is always going to be a limit to state.

> but also works well for what we can do now (ie mostly backfill from live history)
Perhaps, but still wrong IMHO.